onnx/defs/quantization/defs.cc

```diff
@@ -97,7 +104,9 @@ ONNX_OPERATOR_SET_SCHEMA(
             "The type of the input `y_zero_point` and the output `y`.")
         .SetDoc(QuantizeLinear_ver21_doc)
         .TypeAndShapeInferenceFunction([](ONNX_NAMESPACE::InferenceContext& ctx) {
-          if (ctx.hasInput(2)) {
+          if (ctx.getAttribute("output_dtype") != nullptr) {
+            propagateElemTypeFromAttributeToOutput(ctx, "output_dtype", 0);
```

Comment:
nit: would be useful to check, if zeropoint is also specified, that it has the same type.

Reply:
I guess output_dtype is also required to be in the set of allowed T2 types.

Reply:
@gramalingam 

> nit: would be useful to check, if zeropoint is also specified, that it has the same type.

I've added a check that zeropoint and output_dtype match if both are specified. Also added type inference tests to check all cases (only zero-point, only output_dtype, both, both mismatch).

> I guess output_dtype is also required to be in the set of allowed T2 types.

True. I was surprised I couldn't find something similar in any other operator. This should be fixed as part of `onnx/defs/shape_inference.h::propagateElemTypeFromAttributeToOutput`, since this affects other operators (i.e. Cast).
I don't have a quick fix for this, but I've filed #5960 to follow up on that. This PR also includes a couple of skipped tests that should be unskipped once #5960 is fixed.